### ISSUE TITLE
2.6.0 to 3.9.2: a very painful upgrade

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#316]
In the course of upgrade 2.6.0 to 3.9.2 we had several hours of down time for the very first time ever since we began using LizardFS about a year ago.

As instructed in `UPGRADE`, we successfully upgraded shadow master and metalogger (the latter was stopped until upgrade is completed).

However when I restarted _master_, it did not become operational even 30 minutes later (usually restart of master takes less than 5 minutes on 2.6.0). Master was using 100% of one CPU and not responding to connection requests.
It did not stop on `TERM` signal so I had to kill it. Then I promoted already upgraded shadow master which choked in the similar manner and did not respond to connection requests from clients and chunkservers:

    mfschunkserver: connection failed, error: ECONNREFUSED (Connection refused)
    mfsmount: can't connect to master
    mfsmount: master: register error (read header: ENOTCONN (Transport endpoint is not connected))

It seems that master chokes on high volume of information from chunkservers.
Master became available only when I stopped several chunkservers and then re-started 'em one by one. Master became unresponsive every time I started another chunkserver and any client connections (as well as healthy chunkserver connections) were dropped. Chunkservers were complaining (`got unknown message`) and no client could work. Eventually I had to upgrade _all_ chunkservers and re-mount all exports with new clients. Only then cluster became operational again but I had to be careful to avoid starting more than few chunkservers at a time to minimise cluster-wide down time.

Apparently 3.9.2/master is slower than 2.6.0 and it is very noticeable how it struggles to process all the data from chunkservers. If one fat chunkserver is started, others losing connection to master and flood it in attempt to reconnect so it never recovers from race without intervention. 2.6.0 was not vulnerable to this problem.

**Comment by blink69** [comment]
Could you send us more info about your instalation, numbers of chunkservers, client connected. Numbers of inodes, files and directory from metadata.

**Comment by onlyjob** [comment]
7 chunkservers (37 HDDs); ~41_000_000 files; master uses 12 GiB RAM (~25% of total RAM; shadow master uses ~50% RAM); 
Number of clients/mounts reduced to 5 or 6 before upgrade.
Mostly runs on Debian Jessie/amd64; few clients run Debian "testing".

**Comment by blink69** [comment]
Please send us syslog from master, shadows, chunkservers support@lizardfs.com

**Comment by onlyjob** [comment]
That would be hard to do. Things went wrong and I was fiddling for hours until I've managed to get things work. I simply have no time to retrace my steps and collect perfectly synchronised logs from multiple servers... Then I would have to spend even more time interpreting 'em... Don't you think there are enough information already for investigation?
See for yourself what happens when you restart a chunkserver on 3.9.2 -- for some time master just stops responding with 100% CPU utilisation. If you restart several chunkservers at the same time the effect would be worse. More data on chunkserver causes more dramatic effect. Try master 3.9.2 and chunkservers 2.6.0 (can you downgrade a chunkserver?). 

It shouldn't be too hard to do some master profiling to identify where the bottleneck is and what are the critical path computations...

**Comment by onlyjob** [comment]
It is confirmed -- standalone master v3.9.2 is unable to (re-)start when all chunkservers are up. Master uses 100% CPU for some time when chunkserver is joining. While processing data from chunkserver (especially when chunkserver have a lot of data to send) master is not responding to clients (who get timeouts) and to other chunkservers. Eventually connection to chunkservers are dropped so they try to reconnect (re-send their data) and it goes on and on.
Re-starting chinkservers helps because they send data slower due to scanning and master is not getting overloaded.
This is a serious regression from 2.6.0 and it is very easy to reproduce. 
(Eventually I may be able to provide logs but there are just timeouts and reconnections over and over).
I believe master should be responding even when it is processing information from joining chunkserver. Let alone 100% CPU utilisation on master, all clients feel when a chunksever is starting because everything freezes....

**Comment by onlyjob** [comment]
It seems that `TCMALLOC_LIBRARY: /usr/lib/libtcmalloc_minimal.so` (from "libgoogle-perftools-dev") is the culprit. 
I re-built LizardFS-3.9.2 without "libgoogle-perftools-dev" and it had a profound effect on performance of Master.
With "libgoogle-perftools-dev":
~~~~
Oct 30 13:01:19 debstor mfsmaster[19198]: loading objects (files,directories,etc.) from the metadata file
Oct 30 13:01:39 debstor mfsmaster[19198]: loading names from the metadata file
Oct 30 13:02:17 debstor mfsmaster[19198]: loading deletion timestamps from the metadata file
~~~~
(Here master stuck for over 10 minutes with 100% CPU utilisation).

Without "libgoogle-perftools-dev":

~~~~
Oct 30 13:18:32 debstor mfsmaster[20991]: loading objects (files,directories,etc.) from the metadata file
Oct 30 13:18:48 debstor mfsmaster[20991]: loading names from the metadata file
Oct 30 13:19:21 debstor mfsmaster[20991]: loading deletion timestamps from the metadata file
Oct 30 13:19:24 debstor mfsmaster[20991]: loading extra attributes (xattr) from the metadata file
Oct 30 13:19:24 debstor mfsmaster[20991]: loading access control lists from the metadata file
Oct 30 13:19:24 debstor mfsmaster[20991]: loading quota entries from the metadata file
~~~~
(only about 3 seconds for "loading deletion timestamps from the metadata file" comparing to over 10 min. before).

Problem is not completely gone though. Clients are still getting timeouts like the following

    mfsmount[1590]: master: register error (read header: ETIMEDOUT (Operation timed out))

while master is processing data from chunkserver(s).

**Comment by onlyjob** [comment]
Master built without "libgoogle-perftools-dev" (`-DENABLE_TCMALLOC=NO`) is much faster but it is still choking on restart:

~~~~
Nov 01 00:24:09 debstor mfsmaster[18822]: [ OK ] mfsmaster daemon initialized properly
Nov 01 00:24:11 debstor mfschunkserver[25953]: connecting to Master
Nov 01 00:24:11 debstor mfschunkserver[25953]: connected to Master
[...]
Nov 01 00:24:28 debstor mfsmaster[18822]: chunkserver register begin (packet version: 5) - ip: 192.168.0.250, port: 9422
Nov 01 00:25:14 debstor mfsmaster[18822]: chunkserver register end (packet version: 5) - ip: 192.168.0.250, port: 9422, usedspace: 21297288581120 (19834.65 GiB), totalspace: 27594570448896 (25699.45 GiB)
[...]
Nov 01 00:25:15 debstor mfsmaster[18822]: main master server module: (ip:192.168.0.250) write error: EPIPE (Broken pipe)
Nov 01 00:25:15 debstor mfsmaster[18822]: connection with CS(192.168.0.250) has been closed by peer
Nov 01 00:25:15 debstor mfsmaster[18822]: chunkserver disconnected - ip: 192.168.0.250, port: 9622, usedspace: 0 (0.00 GiB), totalspace: 0 (0.00 GiB)
Nov 01 00:25:15 debstor mfsmaster[18822]: chunkserver disconnected - ip: 192.168.0.250, port: 9422, usedspace: 21297288581120 (19834.65 GiB), totalspace: 27594570448896 (25699.45 GiB)
~~~~

So it tries again

~~~~
Nov 01 00:26:00 debstor mfsmaster[18822]: chunkserver register begin (packet version: 5) - ip: 192.168.0.250, port: 9422
Nov 01 00:26:21 debstor mfschunkserver[25953]: connecting to Master
Nov 01 00:26:21 debstor mfschunkserver[25953]: connected to Master
Nov 01 00:26:22 debstor mfsmaster[18822]: connection with CS(192.168.0.250) has been closed by peer
Nov 01 00:26:22 debstor mfsmaster[18822]: chunkserver disconnected - ip: 192.168.0.250, port: 9422, usedspace: 0 (0.00 GiB), totalspace: 0 (0.00 GiB)
Nov 01 00:26:22 debstor mfsmaster[18822]: main master server module: (ip:192.168.0.250) write error: EPIPE (Broken pipe)
Nov 01 00:26:38 debstor mfsmaster[18822]: connection with CS(192.168.0.250) has been closed by peer
Nov 01 00:26:38 debstor mfsmaster[18822]: chunkserver disconnected - ip: 192.168.0.250, port: 9622, usedspace: 0 (0.00 GiB), totalspace: 0 (0.00 GiB)
Nov 01 00:26:38 debstor mfsmaster[18822]: chunkserver register begin (packet version: 5) - ip: 192.168.0.250, port: 9422
Nov 01 00:26:38 debstor mfsmaster[18822]: mfsmaster[18822]: chunkserver register begin (packet version: 5) - ip: 192.168.0.250, port: 9422
Nov 01 00:27:26 debstor mfschunkserver[25953]: connecting to Master
Nov 01 00:27:26 debstor mfschunkserver[25953]: connected to Master
Nov 01 00:27:26 debstor mfsmaster[18822]: main master server module: (ip:192.168.0.250) write error: EPIPE (Broken pipe)
[...]
Nov 01 00:37:28 debstor mfsmaster[18822]: connection with CS(192.168.0.250) has been closed by peer
Nov 01 00:37:28 debstor mfsmaster[18822]: chunkserver disconnected - ip: 192.168.0.250, port: 9622, usedspace: 0 (0.00 GiB), totalspace: 0 (0.00 GiB)
Nov 01 00:37:28 debstor mfsmaster[18822]: main master server module: (ip:192.168.0.250) write error: EPIPE (Broken pipe)
Nov 01 00:37:28 debstor mfsmaster[18822]: chunkserver register begin (packet version: 5) - ip: 192.168.0.250, port: 9422
Nov 01 00:37:28 debstor mfsmaster[18822]: mfsmaster[18822]: chunkserver register begin (packet version: 5) - ip: 192.168.0.250, port: 9422
Nov 01 00:38:16 debstor mfschunkserver[25953]: connecting to Master
Nov 01 00:38:16 debstor mfschunkserver[25953]: connected to Master
[...]
~~~~

and it goes on and on for ~30 minutes until I interrupt vicious cycle by restarting chunkserver:

~~~~
ov 01 00:52:00 debstor mfschunkserver[25953]: terminate signal received
Nov 01 00:52:00 debstor mfschunkserver[25953]: mfschunkserver[25953]: terminate signal received
Nov 01 00:52:00 debstor mfschunkserver[25953]: closing *:9422
Nov 01 00:52:00 debstor mfschunkserver[25953]: mfschunkserver[25953]: closing *:9422
Nov 01 00:52:02 debstor mfsmaster[18822]: connection with CS(192.168.0.250) has been closed by peer
Nov 01 00:52:02 debstor mfsmaster[18822]: chunkserver disconnected - ip: 192.168.0.250, port: 9422, usedspace: 0 (0.00 GiB), totalspace: 0 (0.00 GiB)
Nov 01 00:52:02 debstor mfsmaster[18822]: main master server module: (ip:192.168.0.250) write error: EPIPE (Broken pipe)
Nov 01 00:52:02 debstor mfsmaster[18822]: main master server module: (ip:192.168.0.250) write error: EPIPE (Broken pipe)
[...]
Nov 01 00:52:09 debstor mfsmaster[18822]: main master server module: (ip:192.168.0.250) write error: EPIPE (Broken pipe)
~~~~

As soon as chunkserver is restarted it is joining slowly as it scans its assets. Master's CPU utilisation drops from 100% to ~30%...40% and it begins responding to clients.
Please note that I've picked only log entries for a particular (heaviest) chunkserver. Connections to other chunkservers are dropped as well so (as far as I can tell) it is not important which chunkserver(s) to restart.
From curiosity I tried to restart few other chunkservers instead of this one and managed to achieve the similar results.
This problem makes it impossible to restart Master without restarting some chunkservers hence failover to shadow master is broken/affected as well.

**Comment by psarna** [comment]
Problem recognized. Our issue is that code responsible for counting statistics was significantly changed. When a heavy chunkserver comes up, it sends information about its chunks and they all trigger statistics update, which becomes a bottleneck.
Tcmalloc library seemed to be a culprit because it didn't handle a lot of small allocations well (which is kind of funny, because it's advertised as a great allocator for small objects).

To sum up, sorry for causing problems, we will prepare a fix:)
Tcmalloc would probably stop being our default allocation library as well.

**Comment by onlyjob** [comment]
Is there any relief for this problem? A patch or a build option?
Restarting master is such a hassle on 3.9.2: I have to stop some chunkservers, restart master, wait till it process data from chunkserves, start remaining chunkservers and wait patiently as while chunkservers are joining all clients get timeouts and nobody can work.
The whole process takes at least 20...30 minutes (if done properly) and needs manual intervention.
Merely restarting master (or promoting shadow master)  _always_ fails (never finishes, to be precise) with endless loop of chinkservers joining, losing connection and re-joining again...
This is on LizardFS built without `tcmalloc` as the latter makes it even worse...

**Comment by psarna** [comment]
If everything goes smoothly, you should see patches related to this issue today. It is a priority right now.

**Comment by psarna** [comment]
Well, today's not gonna happen, but we are close. You can track our progress up here: http://cr.skytechnology.pl:8081/#/c/2239/48. Good news is that the preliminary results show that bottleneck is no longer present in master.

**Comment by onlyjob** [comment]
Thanks for update. Keep up the good work and please let me know once finished so I'll test ASAP. :)

**Comment by psarna** [comment]
Here's a freshly baked commit 96ba6f86a036cd8f7e906715aafd7ffa58f74459, it should remove the bottleneck (at least it's what our performance tests say). Please share your experience with it, I hope it helps.

**Comment by onlyjob** [comment]
Sensational improvement, Piotr. So much better. Great work, thank you. IMHO it qualifies for new release.

**Comment by psarna** [comment]
Long live @DarkHaze. New release based on this patch is indeed going to happen soon.

**Comment by eronquillo** [comment]
Hi, are there any updates to this problem?  Ever since I upgraded from 2.6.0 to 3.9.2, it has been a nightmare.  Followed the upgrade instructons.  Now, we have no access to my files, just times out.  I currently use Centos 6.7, one master, one shadow master, 3 metaloggers running on 3 chunk servers.  Thanks!

**Comment by eronquillo** [comment]
Ok, I see 3.9.4 available in the yum repository now.  Going to try it out.

**Comment by onlyjob** [comment]
@eronquillo, please upgrade to 3.9.4 ASAP. Most serious issues has been fixed in 3.9.4.

Besides [Debian](http://debian.org) would probably work better. ;)
